src/nvim/ex_cmds2.c

```diff
@@ -2976,6 +2977,13 @@ static void cmd_source(char_u *fname, exarg_T *eap)
   }
 }
 
+static void cmd_source_buffer(exarg_T *eap)
+{
+  for (linenr_T lnum = eap->line1; lnum <= eap->line2; lnum++) {
```

Comment:
calling `do_cmdline` in a loop is wrong. Rather this needs to  pass a new callback function to `do_cmdline` (like `get_str_line` ) which does a `ml_get` a time.

Reply:
Done.